xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp

```diff
@@ -34,467 +39,897 @@
+#define SETTING_TMR_CHANNEL       "timer.channel"
+#define SETTING_TMR_ANYTIME       "timer.anytime"
+#define SETTING_TMR_DAY           "timer.day"
+#define SETTING_TMR_BEGIN         "timer.begin"
+#define SETTING_TMR_END           "timer.end"
+#define SETTING_TMR_WEEKDAYS      "timer.weekdays"
+#define SETTING_TMR_FIRST_DAY     "timer.firstday"
+#define SETTING_TMR_NEW_EPISODES  "timer.newepisodes"
+#define SETTING_TMR_BEGIN_PRE     "timer.startmargin"
+#define SETTING_TMR_END_POST      "timer.endmargin"
+#define SETTING_TMR_PRIORITY      "timer.priority"
+#define SETTING_TMR_LIFETIME      "timer.lifetime"
+#define SETTING_TMR_DIR           "timer.directory"
+
+#define TYPE_DEP_VISIBI_COND_ID_POSTFIX     "visibi.typedep"
+#define TYPE_DEP_ENABLE_COND_ID_POSTFIX     "enable.typedep"
+#define CHANNEL_DEP_VISIBI_COND_ID_POSTFIX  "visibi.channeldep"
+#define ANYTIME_DEP_ENABLE_COND_ID_POSTFIX  "enable.anytimedep"
+
+#define ENTRY_ANY_CHANNEL (-1)
+
+CGUIDialogPVRTimerSettings::CGUIDialogPVRTimerSettings() :
+  CGUIDialogSettingsManualBase(WINDOW_DIALOG_PVR_TIMER_SETTING, "DialogPVRTimerSettings.xml"),
+  m_bIsRadio(false),
+  m_bIsNewTimer(true),
+  m_bTimerActive(false),
+  m_bAnytime(true),
+  m_bFullTextEpgSearch(true),
+  m_iDay(0),
+  m_iWeekdays(PVR_WEEKDAY_NONE),
+  m_iFirstDay(0),
+  m_bNewEpisodesOnly(false),
+  m_iMarginStart(0),
+  m_iMarginEnd(0),
+  m_iPriority(0),
+  m_iLifetime(0)
 {
   m_loadType = LOAD_EVERY_TIME;
 }
 
+// virtual
+CGUIDialogPVRTimerSettings::~CGUIDialogPVRTimerSettings()
+{
+}
+
 void CGUIDialogPVRTimerSettings::SetTimer(CFileItem *item)
 {
-  m_timerItem         = item;
+  if (item == NULL)
+  {
+    CLog::Log(LOGERROR, "CGUIDialogPVRTimerSettings::SetTimer - No item");
+    return;
+  }
 
-  m_timerItem->GetPVRTimerInfoTag()->StartAsLocalTime().GetAsSystemTime(m_timerStartTime);
-  m_timerItem->GetPVRTimerInfoTag()->EndAsLocalTime().GetAsSystemTime(m_timerEndTime);
-  m_timerStartTimeStr   = m_timerItem->GetPVRTimerInfoTag()->StartAsLocalTime().GetAsLocalizedTime("", false);
-  m_timerEndTimeStr     = m_timerItem->GetPVRTimerInfoTag()->EndAsLocalTime().GetAsLocalizedTime("", false);
+  m_timerInfoTag = item->GetPVRTimerInfoTag();
+
+  if (!m_timerInfoTag)
+  {
+    CLog::Log(LOGERROR, "CGUIDialogPVRTimerSettings::SetTimer - No timer info tag");
+    return;
+  }
+
+  // Copy data we need from tag. Do not modify the tag itself until Save()!
+  m_timerType    = m_timerInfoTag->GetTimerType();
+  m_bIsRadio     = m_timerInfoTag->m_bIsRadio;
+  m_bIsNewTimer  = m_timerInfoTag->m_state == PVR_TIMER_STATE_NEW;
+  m_bAnytime     = m_bIsNewTimer || m_timerInfoTag->IsStartAtAnyTime();
+  m_bTimerActive = m_bIsNewTimer || m_timerInfoTag->IsActive();
+  m_strTitle     = m_timerInfoTag->m_strTitle;
+
+  CDateTime now(CDateTime::GetCurrentDateTime());
+  CDateTime startLocalTime = m_bAnytime ? now : m_timerInfoTag->StartAsLocalTime();
+  CDateTime endLocalTime   = m_bAnytime ? now : m_timerInfoTag->EndAsLocalTime();
```

Comment:
Current time should only be used for manual timers.
Guide based timers should get their start and end time from the epg tag, even if any time is selected, users can unselect it and then the time would be wrong.